blog/_posts/2019-07-14-multithreading.md

```diff
@@ -0,0 +1,526 @@
+He sold all of us on it with some nice animated slides, and it also didn't hurt that
+he was willing to do some of the work.
+The plan was to first develop partr as a standalone C library so it could be tested
+and benchmarked on its own, and then integrate it with the Julia runtime.
+
+After Kiran completed the standalone version of partr, we embarked on a series of
+work sessions including Anton Malakhov (also of Intel) to figure out how to do
+the integration.
+The Julia runtime brings many extra features, such as garbage collection and
+event-based I/O, so this was not entirely straightforward.
+Somewhat disappointingly, though not unusually for a complex software project,
+it took much longer than expected --- nearly two years --- to get the new
+system working reliably.
+A later section of this post will explain some of the internals and difficulties
+involved for the curious.
+But first, let's take it for a spin.
+
+## How to use it
+
+To use Julia with multiple threads, set the `JULIA_NUM_THREADS` environment
+variable:
+
+```
+$ JULIA_NUM_THREADS=4 ./julia
+```
+
+The [Juno IDE][] automatically sets the number of threads based on the number of
+available processor cores, and also provides a graphical interface for changing
+the number of threads, so setting the variable manually is not necessary
+in that environment.
+
+The `Threads` submodule of `Base` houses most of the thread-specific functionality,
+such as querying the number of threads and the ID of the current thread:
+
+```
+julia> Threads.nthreads()
+4
+
+julia> Threads.threadid()
+1
+```
+
+`@threads` loops still work, except now I/O is no problem:
+
+```
+julia> Threads.@threads for i = 1:10
+           println("i = $i on thread $(Threads.threadid())")
+       end
+i = 1 on thread 1
+i = 7 on thread 3
+i = 2 on thread 1
+i = 8 on thread 3
+i = 3 on thread 1
+i = 9 on thread 4
+i = 10 on thread 4
+i = 4 on thread 2
+i = 5 on thread 2
+i = 6 on thread 2
+```
+
+Without further ado, let's try some nested parallelism.
+A perennial favorite example is mergesort, which divides its input in half
+and recursively sorts each half.
+The halves can be sorted independently, yielding a natural opportunity
+for parallelism.
+Here is the code:
+
+```
+# sort the elements of `v` in place, from indices `lo` to `hi` inclusive
+function psort!(v, lo::Int=1, hi::Int=length(v))
+    if lo >= hi                       # 1 or 0 elements; nothing to do
+        return v
+    end
+    if hi - lo < 100000               # below some cutoff, run in serial
+        sort!(view(v, lo:hi), alg = MergeSort)
+        return v
+    end
+
+    mid = (lo+hi)>>>1                 # find the midpoint
+
+    half = @par psort!(v, lo, mid)    # task to sort the lower half; will run
+    psort!(v, mid+1, hi)              # in parallel with the current call sorting
+                                      # the upper half
+    wait(half)                        # wait for the lower half to finish
```

Comment:
I think it'd be fun to use `fetch` here (implementing `sort` instead of `sort!`), as that seems harder to me (relative to what other languages provide and do), and we're already making a copy below

Reply:
```julia
julia> function psort(v, lo::Int=1, hi::Int=length(v))
                   if lo >= hi                       # 1 or 0 elements; nothing to do
                       return v
                   end
                   if hi - lo < 100000               # below some cutoff, run in serial
                       return sort(view(v, lo:hi), alg = MergeSort)
                   end
               
                   mid = (lo+hi)>>>1                 # find the midpoint
               
                   half = @task psort(v, lo, mid)    # task to sort the lower half; will run
                   half.sticky = false
                   schedule(half)
                   right = psort(v, mid+1, hi)       # in parallel with the current call sorting
                                                     # the upper half
                   left = fetch(half)                # wait for the lower half to finish
                   out = similar(v, hi-lo+1)         # result
                   @assert length(right) + length(left) == length(out)
               
                   i, il, ir = 1, 1, 1               # merge the two sorted sub-arrays
                   @inbounds while il <= length(left) && ir <= length(right)
                       l, r = left[il], right[ir]
                       if l < r
                           out[i] = l
                           il += 1
                       else
                           out[i] = r
                           ir += 1
                       end
                       i += 1
                   end
                   @inbounds while il <= length(left)
                       out[i] = left[il]
                       il += 1
                       i += 1
                   end
                   @inbounds while ir <= length(right)
                       out[i] = right[ir]
                       ir += 1
                       i += 1
                   end
                   return out
               end
```
```
julia> using Random; Random.seed!(0); a = rand(20000000);

julia> @time sort(a);
  1.469319 seconds (6 allocations: 152.588 MiB)

julia> @time sort(a);
  1.540864 seconds (6 allocations: 152.588 MiB, 2.91% gc time)

julia> @time psort(a); # 1 thead
 20.526943 seconds (879.42 M allocations: 14.520 GiB, 9.55% gc time)

julia> @time psort(a); # 2 threads
 12.870170 seconds (879.42 M allocations: 14.520 GiB, 15.73% gc time)

julia> @time psort(a);
 10.782067 seconds (879.42 M allocations: 14.520 GiB, 18.62% gc time)

julia> @time psort(a); # 4 threads
  9.499449 seconds (879.42 M allocations: 14.520 GiB, 22.32% gc time)
```
```
      ┌────────────────────────────────────────┐ 
   30 │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
      │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
      │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
      │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
      │⢄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
      │⠀⠉⠒⠤⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
      │⠀⠀⠀⠀⠈⠑⠢⢄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
      │⠀⠀⠀⠀⠀⠀⠀⠀⠉⠒⢄⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
      │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⠑⠦⠤⢄⣀⣀⣀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
      │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠉⠉⠉⠒⠒⠒⠤⠤⠤⠤⢄⣀⣀⣀⣀⣀⠀⠀⠀⠀⠀│ 
      │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠉⠉⠉⠉⠉│ 
      │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
      │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
      │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
    0 │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
      └────────────────────────────────────────┘ 
      1                                        4
```